makedef.pl: Eliminate need for embed.fnc sync

khwilliamson opened the pull request:
makedef.pl has to manually be kept in sync with various other pieces of code.  This is extra work, and error prone.  This commit eliminates that need for embed.fnc syncing.

The relatively new HeaderParser already returns the information we need in its object for each entry.  This commit just takes advantage of that, and removes the many lines that were previously needed for this purpose.

Running this under Linux with PLATFORM=test revealed nearly 50 symbols that would have been needlessly exported, such as Perl_grok_bslash_x(), which is only defined in certain files, and not globally.


* This set of changes does not require a perldelta entry.

khwilliamson committed:
makedef.pl: Refactor $variable initialization

This is so the string doesn't have to be repeated

khwilliamson committed:
makedef.pl: Use a name for $_

A future commit will refer to this further below; better to have a name
than $_ for that.

khwilliamson committed:
makedef.pl: Remove nonsensical comment

The 'o' flag requires a Perl_prefix, so shouldn't skip it.  Perhaps it
was a typo, meant to be the 'O' flag.  But that flag now also requires
the p flags, so also a Perl_prefix.

khwilliamson committed:
embed.fnc: mv definition of more_sv

This was in a #ifdef of being in sv.c, which it is, but since it is
public, it needs to be moved out of this.  This removes the need for a
copy of its prototype to be in sv_inline.h

khwilliamson committed:
makedef.pl: Eliminate need for embed.fnc sync

makedef.pl has to manually be kept in sync with various other pieces of
code.  It has to know the conditionals that various symbols are compiled
under.  This is extra work, and error prone.  This commit eliminates
that need for embed.fnc syncing.

The relatively new HeaderParser already returns the information about
what conditionals are in effect that we need in its object for each
entry.  This commit just takes advantage of that, and removes the many
lines that were previously needed for this purpose.

Running this under Linux with PLATFORM=test revealed about a dozen
symbols that would have been needlessly exported, such as
Perl_grok_bslash_x(), which is only defined in certain files, and not
globally.

khwilliamson committed:
Add some duplication to makedef.pl from perl.h

There is unfortunately some duplication required in makedef.pl copied
from perl.h.  This adds some more for two variables, required for the
next commit, which will remove far more lines from makedef.pl than this
adds

tonycoz commented:
`iAdd some duplication to makedef.pl from perl.h `

typo (extra `i`) in commit message.

khwilliamson committed:
regen/embed.pl: Fix typos in comment

A misspelling and a run-on sentence

khwilliamson committed:
regen/embed.pl: Add checks that O flag used correctly

And add some comments in embed.fnc

khwilliamson committed:
regen/embed.pl: Avoid use of hard-coded list

The list consists of exactly the functions that have the O flag set in
embed.fnc.  No need to keep this data twice.  The entries are trivially
generatable from existing entries as we go along

And those generated entries have the added advantage of not using the
short name, so potentially less name space pollution

tonycoz commented on regen/embed.pl:
```diff
@@ -846,9 +846,9 @@ sub embed_h {
-                    # uncondtionally because existing code might call e.g.
-                    # warn() without aTHX in scope.
+                    # PERL_CORE or PERL_WANT_VARARGS.  These should not be
+                    # done uncondtionally because existing code might call
```
could fix "uncond??tionally" while we're at it.
